### PR TITLE
[FunctionComparator] Differentiate instructions passing different MDStrings

### PR DESCRIPTION
**Pull request by NuriAmari**
Prior to this patch, differing metadata operands to two otherwise identical instructions was not enough to consider the instructions different in the eyes of the function comparator. This breaks LLVM virtual function elimination, among other features.

In this patch, we handle the case where two associated operands are MDStrings of different value. This patch does not differentiate more complex metadata operands.

### CODE REVIEW

**Comment by NuriAmari** [comment]
There are a couple related patches: https://github.com/llvm/llvm-project/pull/65963, https://github.com/llvm/llvm-project/commit/8325d46a4ab803c6eeff28bf1bdafb43287ce557, but they don't solve this problem.

File: llvm/test/Transforms/MergeFunc/mergefunc-preserve-vfe-intrinsics.ll [hunk 1]
```diff
@@ -0,0 +1,34 @@
+; This test contains three identical functions, aside from the metadata
+; they pass to a function call. This test verifies that the function merger
+; pass is able to merge the two functions that are truly identifical,
```
Comment (drodriguez) [inline llvm/test/Transforms/MergeFunc/mergefunc-preserve-vfe-intrinsics.ll:3]: Nit: typo `identifical` → `identical`

File: llvm/test/Transforms/MergeFunc/mergefunc-preserve-vfe-intrinsics.ll [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+; but the third that passes different metadata is preserved
+
+; RUN: opt -passes=mergefunc -S %s | FileCheck %s
```
Comment (nikic) [inline llvm/test/Transforms/MergeFunc/mergefunc-preserve-vfe-intrinsics.ll:6]: Use update_test_checks.py please, and move the test description below the RUN line.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by dexonsmith** [comment]
Can you change the subject of the commit / pull request before pushing to avoid using the phrase "distinct MDString"? I suggest, "different" instead of "distinct". The problem is that "distinct" means something specific in metadata: non-uniqued. Although this cannot be applied to `MDString`, it's still a bit confusing.

**Review by dexonsmith — DISMISSED** [review]

File: llvm/lib/Transforms/Utils/FunctionComparator.cpp [hunk 3]
```diff
@@ -160,10 +160,23 @@ int FunctionComparator::cmpAttrs(const AttributeList L,
...
   // It ignores any other cases, so that the metadata nodes are considered
   // equal even though this is not correct.
   // We should structurally compare the metadata nodes to be perfect here.
+
+  auto *MDStringL = dyn_cast<MDString>(L);
+  auto *MDStringR = dyn_cast<MDString>(R);
+  if (MDStringL && MDStringR) {
+    if (MDStringL == MDStringR)
+      return 0;
+    return MDStringL->getString().compare(MDStringR->getString());
```
[the remark this answers is not in the document]
Comment (dexonsmith) [reply llvm/lib/Transforms/Utils/FunctionComparator.cpp:173]: Oh, sorry, I misread this, never mind. LG with the subject change I already requested. Give me a moment to figure out how to update the PR....

**Review by dexonsmith — APPROVED** [review]
LGTM, if you rephrase the subject of the commit and pull request to use "different" instead of "distinct" when referring to `MDString` (because of the language clash against distinct vs. uniqued in LLVM metadata).

**Comment by NuriAmari** [comment]
Thanks for the reviews all!

**Pull request merged**